src/main/java/bio/terra/workspace/db/ResourceDao.java

```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
         .collect(Collectors.toList());
   }
 
+  /**
+   * @param workspaceId identifier for work space to enumerate
+   * @param controlledResourceIds identifiers of controlled resources visible to the caller
+   * @param resourceType filter by this resource type - optional
+   * @param stewardshipType filtered by this stewardship type - optional
+   * @param offset starting row for result
+   * @param limit maximum number of rows to return
+   * @return list of resources
+   */
+  @Transactional(
+      propagation = Propagation.REQUIRED,
+      isolation = Isolation.SERIALIZABLE,
+      readOnly = true)
+  public List<WsmResource> enumerateResources(
+      UUID workspaceId,
+      List<UUID> controlledResourceIds,
+      @Nullable WsmResourceType resourceType,
+      @Nullable StewardshipType stewardshipType,
+      int offset,
+      int limit) {
+    List<UUID> idList = Optional.ofNullable(controlledResourceIds).orElse(new ArrayList<>());
+
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue("workspace_id", workspaceId.toString())
+            .addValue("offset", offset)
+            .addValue("limit", limit);
+
+    StringBuilder sb = new StringBuilder(RESOURCE_SELECT_SQL);
+    if (resourceType != null) {
+      sb.append(" AND resource_type = :resource_type");
+      params.addValue("resource_type", resourceType.toSql());
+    }
+
+    // There are three cases for the stewardship type filter
+    // 1. If it is REFERENCED, then we ignore idList and just filter
+    //    for referenced resources.
+    // 2. If it is CONTROLLED, and the idList is not empty, then we filter for
+    //    CONTROLLED and require that the resources be in the idList.
+    // 3. If it is null, then we want the OR of 1 and 2
+    // Note that we supply the constant value as a parameter so that the SQL form
+    // is defined in exactly one place.
+    boolean includeReferenced = (stewardshipType == null || stewardshipType == REFERENCED);
+    boolean includeControlled =
+        (!idList.isEmpty() && (stewardshipType == null || (stewardshipType == CONTROLLED)));
+
+    final String referencedPhrase = "stewardship_type = :referenced_resource";
+    final String controlledPhrase =
+        "(stewardship_type = :controlled_resource AND resource_id IN (:id_list))";
+
+    sb.append(" AND ");
+    if (includeReferenced && includeControlled) {
```

Comment:
If we get any fancier, we should look into dedicated query building classes/DSLs.